File: graphs/210.java
import java.util.*;

class Solution {  // BFS topologySort, adjacenyMap + indices list,
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adjList = new ArrayList<>();
        int[] indices = new int[numCourses];
        for (int i = 0; i < numCourses; i++){
            adjList.add(new LinkedList<>());
        }
        for (var pr : prerequisites){
            int to = pr[0];
            int from = pr[1];
            adjList.get(from).add(to);
            indices[to]++;
        }

        Queue<Integer> q = new LinkedList<>();
        for (int c = 0; c < numCourses; c++){
            if (indices[c] == 0){
                q.offer(c);
            }
        }

        int[] res = new int[numCourses];
        int resInd = 0;
        while (!q.isEmpty()){
            int course = q.poll();
            res[resInd] = course;
            resInd++;

            for (int to : adjList.get(course)){
                indices[to]--;
                if (indices[to] == 0){
                    q.offer(to);
                }
            }
        }

        return resInd == numCourses ? res : new int[]{};
    }
}

class SolutionDFS {  // DFS topologySort, only adjacencyMap
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> adjMap = new HashMap<>();

        for (int[] pr : prerequisites) {
            adjMap.computeIfAbsent(pr[1], k -> new LinkedList<>())
                    .add(pr[0]);
        }

        BitSet visited = new BitSet(numCourses);
        BitSet path = new BitSet(numCourses);
        Deque<Integer> sortStack = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (!visited.get(i)) {
                var order = topSort(i, adjMap, visited, sortStack, path);
                if (order == null) {
                    return new int[]{};
                }
            }
        }

        int[] res = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            res[i] = sortStack.pop();
        }
        return res;
    }

    public Deque<Integer> topSort(int course, Map<Integer, List<Integer>> adjMap, BitSet visited, Deque<Integer> order, BitSet path) {
        if (path.get(course)) {
            return null;
        }

        if (visited.get(course)) {
            return order;
        }

        visited.set(course);
        path.set(course);

        for (int neigh : adjMap.getOrDefault(course, new LinkedList<>())) {
            var neigOrder = topSort(neigh, adjMap, visited, order, path);
            if (neigOrder == null) {
                return null;
            }
        }

        path.clear(course);

        order.push(course);

        return order;
    }
}

File: problems/310/medium/_py/210.py
from collections import deque
from typing import List


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        outdegree = {}
        indegree = {}

        for pair in prerequisites:
            outs = outdegree.setdefault(pair[1], set())
            outs.add(pair[0])
            ins = indegree.setdefault(pair[0], set())
            ins.add(pair[1])

        q = deque()
        for course in range(numCourses):
            if course not in indegree:
                q.append(course)

        res = []
        while q:
            course = q.popleft()
            res.append(course)
            for nextCourse in outdegree.get(course, set()):
                indegree[nextCourse].remove(course)
                if len(indegree[nextCourse]) == 0:
                    indegree.pop(nextCourse)
                    q.append(nextCourse)

        return res if len(res) == numCourses else []


class SolutionIndegreeAsAList:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adjMap = {}
        indegree = [0]*numCourses

        for pair in prerequisites:
            adjMap.setdefault(pair[1], set()).add(pair[0])
            indegree[pair[0]] += 1

        q = deque()
        for course in range(numCourses):
            if indegree[course] == 0:
                q.append(course)

        res = []
        while q:
            course = q.popleft()
            res.append(course)
            for nextCourse in adjMap.get(course, set()):
                indegree[nextCourse] -= 1
                if indegree[nextCourse] == 0:
                    q.append(nextCourse)

        return res if len(res) == numCourses else []

